File: 0131 Palindrome Partitioning/solution.py
#! python3

class Solution:
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        self.res = []
        self.dfs(s, [])
        return self.res
    
    def dfs(self, s, cur):
        if not s:
            self.res.append(cur[:])
        
        for i in range(1, len(s)+1):
            if self.isPali(s[:i]):
                cur.append(s[:i])
                self.dfs(s[i:], cur)
                cur.pop()
            
    def isPali(self, s):
        return s == s[::-1]


if __name__ == "__main__":
    print(Solution().partition("aab"))

File: 0131 Palindrome Partitioning/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<string>> res;
    vector<string> cur;

    vector<vector<string>> partition(string s) {
        dfs(s);
        return res;
    }

    void dfs(string s)
    {
        if (s.empty())
            res.push_back(cur);
        for (int i = 1; i <= s.size(); i ++)
        {
            if (check(s.substr(0, i)))
            {
                cur.push_back(s.substr(0, i));
                dfs(s.substr(i));
                cur.pop_back();
            }
        }
    }

    bool check(string a)
    {
        for (int i = 0, j = a.size() - 1; i < j; i ++, j --)
        {
            if (a[i] != a[j])
                return false;
        }
        return true;
    }
};

int main(){
    string s = "aab";
    vector<vector<string>> res = Solution().partition(s);
    for (auto v : res)
    {
        for (auto s : v)
            cout << s << " ";
        cout << endl;
    }

    return 0;
}